EIPS/eip-signTypedData.md

```diff
@@ -0,0 +1,439 @@
+
+**Definition**: A *member type* can be either an atomic type, a dynamic type or a reference type.
+
+**Definition**: The *atomic types* are `bytes1` to `bytes32`, `uint8` to `uint256`, `int8` to `int256`, `bool` and `address`. These correspond to their definition in Solidity. Note that there are no aliases `uint` and `int`. Note that contract addresses are always plain `address`. Fixed point numbers are not supported by the standard. Future versions of this standard may add new atomic types.
+
+**Definition**: The *dynamic types* are `bytes` and `string`. These are like the atomic types for the purposed of type declaration, but their treatment in encoding is different.
+
+**Definition**: The *reference types* are arrays and structs. Arrays are either fixed size or dynamic and denoted by `Type[n]`  or `Type[]` respectively. Structs are references to other structs by their name. The standard supports recursive struct types.
+
+**Definition**: The set of structured typed data `𝕊` contains all the instances of all the struct types.
+
+### Definition of `hashStruct`
+
+The `hashStruct` function is defined as
+
+* `hashStruct(s : 𝕊) = keccak256(typeHash ‖ encodeData(s))` where `typeHash = keccak256(encodeType(typeOf(s)))`
+
+**Note**: The `typeHash` is a constant for a given struct type and does not need to be runtime computed.
+
+### Definition of `encodeType`
+
+The type of a struct is encoded as `name ‖ "(" ‖ member₁ ‖ "," ‖ member₂ ‖ "," ‖ … ‖ memberₙ ")"` where each member is written as `type ‖ " " ‖ name`. For example, the above `Message` struct is encoded as `Message(address from,address to,string contents)`.
+
+If the struct type references other struct types (and these in turn reference even more struct types), then the set of referenced struct types is collected, sorted by name and appended to the encoding. An example encoding is `Transaction(Person from,Person to,Asset tx)Person(address wallet,string name)Asset(address token,uint256 amount)`.
+
+### Definition of `encodeData`
+
+The encoding of a struct instance is `value₁ ‖ value₂ ‖ … ‖ valueₙ`, i.e. a simple concatenation of the encoding of the member values. Each encoded member value is exactly 32-byte long.
+
+The atomic values are encoded as follows: Boolean `false` and `true` are encoded as `uint256` values `0` and `1` respectively. Addresses are encoded as `uint160`. Integer values are sign-extended to 256-bit and encoded in big endian order. `bytes1` to `bytes31` are padded at the end to `bytes32`.
+
+The dynamic values `bytes` and `string` are encoded as a `keccak256` hash of their contents.
+
+The array values are encoded as the `keccak256` hash of the concatenated `encodeData` of their contents (i.e. the encoding of `SomeType[5]` identical to that of a struct containing five members of type `SomeType`).
+
+The struct values are encoded recursively as `hashStruct(value)`. This is undefined for cyclical data.
+
+### Extension of Solidity
```

Comment:
This should probably be standardised separately; you can't enforce changes to Solidity in an EIP. :)

Reply:
Removed it. What is the recommended process here, file an issue or PR an implementation to the Solidity repo after the EIP is accepted?

Reply:
The usual process for solidity is to create an issue.

Note this would add a member to structs for a functionality that is external to solidity, something we have not done before. Also, this document only concerns a subset of the valid types of Solidity. Because of that, the member cannot exist on all possible structs. It might also be an option to just provide access to a string representation of the type as defined here and apply the hash function seperately, which would potentially give higher flexibility.

Reply:
> Note this would add a member to structs for a functionality that is external to solidity, something we have not done before.

IMHO, I see it on similar footing to how the ABI has `abi.encode`, `someFunction.selector` and the special case for `bytes4` as the first argument in `call`. In those cases Solidity adds an intrinsic to the language to facility a standard external to the language. But the ABI is of course more fundamental to Ethereum than this modest proposal.

Another way to look at it is to view it as an incremental improvement in the `keccak256` function. Currently the behaviour when hashing arguments of a struct type is to hash the storage location (as opposed to the contents). I would argue that the current behaviour is counter intuitive and potentially dangerous (see for reference the insecure proposals applying `keccak256` to a struct in this PRs comment thread). Replacing the current behaviour for `keccak256` on struct types with the one proposed in this EIP would IMHO improve Solidity as a language, irregardless of whether this has applications outside of Solidity.

> Also, this document only concerns a subset of the valid types of Solidity. Because of that, the member cannot exist on all possible structs.

Correct. I did design it with Solidity in mind and I believe there is a straightforward extension of the standard that would cover all Solidity structs while remaining compatible on the subset that overlaps.

> It might also be an option to just provide access to a string representation of the type as defined here and apply the hash function seperately, which would potentially give higher flexibility.

The main reason I would like intrinsic support is to prevent programmer error when implementing the standard. Both the (hash of) the string representation of the type and the encoding for the hash function itself are easily derived from the type info, but would otherwise have to be manually maintained. An intrinsic would also make it easier to use the optimized version.